www/docs/server/infer-types.md

```diff
@@ -5,57 +5,101 @@ sidebar_label: Inferring Types
-
-- `inferProcedureOutput<TProcedure>`
-- `inferProcedureInput<TProcedure>`
+```twoslash include server
+// @module: esnext
+// @filename: server.ts
+import { initTRPC } from '@trpc/server';
+import { z } from "zod";
 
-Let's assume we have this example router:
+const t = initTRPC()();
 
-```ts title='routers/_app.ts'
-export const appRouter = t.router({
+const appRouter = t.router({
   user: t.router({
     me: t.procedure.query(() => {
-      // ...
+      // imaginary query for current user
       return { id: 1, name: 'John' };
     }),
   }),
-  post: t.procedure.input(z.string()).mutation(() => {
-    // ...
-    return post.id;
+  postCreate: t.procedure.input(z.string())
+    .mutation(({ input }) => {
+      // imaginary db call
+      return { id: 1, title: 'tRPC is the best!' };
     }),
 });
 
 export type AppRouter = typeof appRouter;
 ```
 
+It is often useful to wrap functionality of your `@trpc/client` or `@trpc/react` api within other functions. For this purpose, it's necessary to be able to infer input types and output types generated by your `@trpc/server` router.
+
+## Inference Helpers
+
+`@trpc/server` exports the following helper types to assist with inferring these types from the `AppRouter` exported by your `@trpc/server` router:
+
+- `inferProcedureOutput<TProcedure>`
+- `inferProcedureInput<TProcedure>`
+
+Let's assume we have this example router:
+
+```ts twoslash title='server.ts'
+// @include: server
+```
+
 By traversing the router object, you can infer the types of the procedures. The following example shows how to infer the types of the procedures of the example `appRouter`:
 
-```ts title='types.ts'
-import type { 
-  inferProcedureOutput, 
-  inferProcedureInput, 
-} from '@trpc/server';
-import type { AppRouter } from 'routers/_app';
+```ts twoslash title="client.ts"
+// @module: esnext
+// @include: server
+// ---cut---
+// @filename: client.ts
+import type { inferProcedureInput, inferProcedureOutput } from '@trpc/server';
+import type { AppRouter } from './server';
+
+// @noErrors
+type UserOutput = inferProcedureOutput<AppRouter['user']['me']>;
+//   ^?
+type PostInput = inferProcedureInput<AppRouter['postCreate']>;
+//   ^?
+```
+
+### Additional DX Helper Type
+
+if you don't like the double-import of first the inferance helper, then your `AppRouter` everytime you want to infer a type, `@trpc/server` also exports a type `GetInferanceHelpers<TRouter>`. This lets you pass your router at initialization then import a single helper type when inferring types:
 
-export type User = inferProcedureOutput<AppRouter['user']['me']>;
-//           ^? { id: number, name: string }}
+```ts twoslash title='utils/trpc.ts'
+// @include: server
+// ---cut---
+import type { GetInferenceHelpers } from '@trpc/server';
+import type { AppRouter } from './server';
 
-export type PostInput = inferProcedureInput<AppRouter['post']>;
-//           ^? string
+// @noErrors
+export type InferTRPC = GetInferenceHelpers<AppRouter>;
```

Comment:
Maybe call it some more intuitive? Can't think of a good name tho

Reply:
Would it help if we modify the helper-type and do something like this to keep a similar API as the current one. Two imports but same `utils`-path as the rest which still solves the issue imo:

```ts
// utils/trpc.ts
import type { GetInferenceHelper } from "@trpc/server";
import type { AppRouter } from "./server";

export type InferProcedureInput = GetInferenceHelper<AppRouter, "input">;
export type InferProcedureOutput = GetInferenceHelper<AppRouter, "output">;

// usage
import { InferProcedureInput } from "~/utils/trpc";
type Post = InferProcedureOutput['post']['byId'];
```

Reply:
Pushed that one up so you can see it in action. 👍🏼 or 👎🏼